### ISSUE TITLE
Loss spike after checkpoint reload

### ISSUE THREAD

**Issue by OverLordGoldDragon** [PyTorchLightning/pytorch-lightning#10766]
[Duplicate](https://github.com/NVIDIA/apex/issues/480), supposedly fixed in PL, but I reproduce it with PL 1.5.3, torch 1.10, cudatoolkit 11.3.1, Windows 10. My spikes are drastically worse (loss = 30) despite using generous warmup with lr starting at 1e-8. I use `precision=16`, can provide MRE if needed.

What's the status? Is there a workaround?

cc @tchaton @carmocca @awaelchli @ninginthecloud @jjenniferdai

**Comment by tchaton** [comment]
Dear @OverLordGoldDragon,

Would you mind providing a reproducible script? I believe it is possible the scheduler isn't properly restarted. Mind using the LrRateMonitor to plot the learning rate?

**Comment by OverLordGoldDragon** [comment]
Scheduler's fine according to wandb and tensorboard,

<img src="https://user-images.githubusercontent.com/16495490/143613542-0634c74f-c794-42d1-9bf3-68025e4610dc.png" width="400">

I noticed in AMP16 there's an initial loss scaling that's very high (`2**16`), and it means to skip if gradients are inf. However this logic is unclear to me as we can get close to overflow but not overflow, so if the intent is that excess gradients shouldn't apply, it's very unreliable. It might also be that upon checkpoint restart, the adapted scaling isn't loaded and it uses initial scaling, hence the spike.

I'll check further and provide MRE.

**Comment by OverLordGoldDragon** [comment]
> if the intent is that excess gradients shouldn't apply

In line with this, the first iteration has a reasonable loss, but the one immediately after produces a spike. This is also (and especially) the case at start of training, no restarts.

**Comment by tchaton** [comment]
Hey @OverLordGoldDragon,

Mind trying `precision="bf16"`. I would need to check but it is possible the scale state dict isn't properly collected and restored.

**Comment by tchaton** [comment]
Dear @OverLordGoldDragon,

Any chance you can provide a reproducible script ?

Best,
T.C

**Comment by OverLordGoldDragon** [comment]
@tchaton I'm no longer sure this is an issue, there were other problems which after resolving made spikes disappear at start of training, but I've yet to reload a checkpoint. I'll check sometime.

**Comment by tchaton** [comment]
Hey @OverLordGoldDragon,

Thanks for the update. Let's close this issue and re-open it if you are sure there is an issue.

**Comment by OverLordGoldDragon** [comment]
Following up, the initial loss spike was due to following the "scheduler after step" advice, which is faulty if scheduler includes warmup; if `self.learning_rate` applies upon checkpoint load rather than the adjusted lr (unless scheduler is called before step), then this also explains the reload spikes.
